src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -5012,5 +5012,782 @@ public async System.Collections.Generic.IAsyncEnumerator<int> GetAsyncEnumerator
+                //     static void M([DefaultCancellation] CancellationToken token)
+                Diagnostic(ErrorCode.WRN_UnconsumedDefaultCancellationAttributeUsage, "DefaultCancellation").WithArguments("token").WithLocation(6, 20)
+                );
+        }
+
+        [Fact, WorkItem(34407, "https://github.com/dotnet/roslyn/issues/34407")]
+        public void CancellationTokenParameter_Indexer()
+        {
+            string source = @"
+using System.Runtime.CompilerServices;
+using System.Threading;
+class C
+{
+    int this[[DefaultCancellation] CancellationToken key] => 0;
+}";
+            var comp = CreateCompilationWithAsyncIterator(new[] { source, DefaultCancellationAttributeType });
+            comp.VerifyDiagnostics(
+                // (6,15): warning CS8424: The DefaultCancellationAttribute applied to parameter 'key' will have no effect. The attribute is only effective on a parameter of type CancellationToken in an async-enumerable method
+                //     int this[[DefaultCancellation] CancellationToken key] => 0;
+                Diagnostic(ErrorCode.WRN_UnconsumedDefaultCancellationAttributeUsage, "DefaultCancellation").WithArguments("key").WithLocation(6, 15)
+                );
+        }
+
+        [Fact, WorkItem(34407, "https://github.com/dotnet/roslyn/issues/34407")]
+        public void CancellationTokenParameter_TwoParameterHaveAttribute()
+        {
+            string source = @"
+using static System.Console;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+class C
+{
+    static async Task Main()
+    {
+        using CancellationTokenSource source = new CancellationTokenSource();
+        CancellationToken token = source.Token;
+        var enumerable = Iter(42, default, default);
+        await using var enumerator = enumerable.GetAsyncEnumerator(token); // some token passed
+
+        if (!await enumerator.MoveNextAsync()) throw null;
+        System.Console.Write($""{enumerator.Current} ""); // 42
+
+        if (!await enumerator.MoveNextAsync()) throw null;
+        System.Console.Write($""{enumerator.Current} ""); // 43
+
+        source.Cancel();
+        try
+        {
+            await enumerator.MoveNextAsync();
+        }
+        catch (System.OperationCanceledException)
+        {
+            Write(""Cancelled"");
+        }
+    }
+    static async System.Collections.Generic.IAsyncEnumerable<int> Iter(int value, [DefaultCancellation] CancellationToken token1, [DefaultCancellation] CancellationToken token2)
+    {
+        yield return value++;
+        await Task.Yield();
+        yield return value++;
+        if (token1.IsCancellationRequested) Write(""token1 cancelled "");
+        if (token2.IsCancellationRequested) Write(""token2 cancelled "");
+        token2.ThrowIfCancellationRequested();
+        Write(""SKIPPED"");
+        yield return value++;
+    }
+}";
+            // TODO2 should we produce a warning here?
```

Comment:
>TODO2 [](start = 15, length = 5)

:memo: I'll convert this to a follow-up issue

Reply:
Please do so before checking in.

---
In reply to: [276848171](https://github.com/dotnet/roslyn/pull/35121#discussion_r276848171) [](ancestors = 276848171)